openedx/features/enterprise_support/enrollments/utils.py

```diff
@@ -38,32 +95,73 @@ def lms_enroll_user_in_course(
+                    or current_enrollment['is_active'] != is_active
+                ):
+                    response = enrollment_api.update_enrollment(
+                        username,
+                        str(course_id),
+                        mode=desired_mode,
+                        is_active=is_active,
+                        enrollment_attributes=None,
+                    )
+                    if not response or (
+                        response['mode'] != desired_mode or
+                        response['is_active'] != is_active
+                    ):
+                        log.exception(
+                            "An error occurred while updating the new course enrollment for user "
+                            "[%s] in course run [%s]",
+                            username,
+                            course_id,
+                        )
+                        raise EnrollmentNotUpdatableError(
+                            f"Unable to upgrade enrollment for user {username} "
+                            "in course {course_id} to {desired_mode} mode."
+                        )
+                return None
+            else:
+                response = enrollment_api.add_enrollment(
+                    username,
+                    str(course_id),
+                    mode=desired_mode,
+                    is_active=is_active,
+                    enrollment_attributes=None,
+                    enterprise_uuid=enterprise_uuid,
+                )
+                if not response:
+                    log.exception(
+                        "An error occurred while creating the new course enrollment for user "
+                        "[%s] in course run [%s]",
+                        username,
+                        course_id,
+                    )
+                    raise CourseEnrollmentError(
+                        f"Unable to create enrollment for user {username} in course {course_id}."
+                    )
+        except CourseEnrollmentExistsError:
+            log.info(
+                "The enrollment for user [%s] in course run [%s] already exists.",
                 username,
-                str(course_id),
-                mode=mode,
-                is_active=is_active,
-                enrollment_attributes=None,
-                enterprise_uuid=enterprise_uuid,
+                course_id,
             )
-            log.info('The user [%s] has been enrolled in course run [%s].', username, course_id)
-            return response
-        except CourseEnrollmentExistsError as error:  # pylint: disable=unused-variable
-            log.warning('An enrollment already exists for user [%s] in course run [%s].', username, course_id)
             return None
-        except CourseEnrollmentError as error:
+        except (CourseEnrollmentError, EnrollmentNotUpdatableError) as error:
             log.exception("An error occurred while creating the new course enrollment for user "
```

Comment:
nit: is this error message slightly misleading now that it could also trigger for upgrading an existing enrollment _and_ the "creating the new course enrollment"?